connaisseur/validators/cosign/cosign_validator.py

```diff
@@ -21,44 +22,110 @@
 class CosignValidator(ValidatorInterface):
     name: str
     trust_roots: list
+    vals: dict
     k8s_keychain: bool
 
     def __init__(self, name: str, trust_roots: list, auth: dict = None, **kwargs):
         super().__init__(name, **kwargs)
         self.trust_roots = trust_roots
+        self.vals = {}
         self.k8s_keychain = False if auth is None else auth.get("k8s_keychain", False)
 
-    def __get_key(self, key_name: str = None):
-        key_name = key_name or "default"
-        try:
-            key = next(
-                key["key"] for key in self.trust_roots if key["name"] == key_name
-            )
-        except StopIteration as err:
-            msg = 'Trust root "{key_name}" not configured for validator "{validator_name}".'
-            raise NotFoundException(
-                message=msg, key_name=key_name, validator_name=self.name
-            ) from err
-        return Key("".join(key))
-
     async def validate(
         self, image: Image, trust_root: str = None, **kwargs
     ):  # pylint: disable=arguments-differ
-        key = self.__get_key(trust_root)
-        return self.__get_cosign_validated_digests(image, key).pop()
 
-    def __get_cosign_validated_digests(self, image: Image, key: Key):
+        required = kwargs.get("required", [])
+        # if not configured, `threshold` is 1 if trust root is not "*" or
+        # `required` is specified and number of trust roots otherwise
+        threshold = kwargs.get(
+            "threshold",
+            (int(any(required)) or len(self.trust_roots)) if trust_root == "*" else 1,
+        )
+
+        self.vals = self.__get_pinned_keys(trust_root, required)
+
+        tasks = [self.__validation_task(k, image) for k in self.vals.keys()]
```

Comment:
You could also do:
```python
tasks = map(lambda k: self.__validation_task(k, image), self.vals.keys())
```

This returns an iterator, preventing in-memory construction of the list.

Now imagine you define a function `self.__validation_task_by_image_callback(image)` that returns a function (!) that only takes `k` as an argument, but returns `self.__validation_task(k, image)`:

```python
def __validation_task_by_image_callback(image):
    return lambda k: self.__validation_task(k, image)  # not sure if this syntax works; think of it as pseudocode
```

Essentially this is the same function, but it has the `image` argument statically bound.

... then you could simply do:

```python
validation_callback = self.__validation_task_by_image_callback(image)
tasks = map(validation_callback, self.vals.keys())
```

This would very clearly show that each item of `keys()` is mapped onto a function call.